// stylelint/dist/rules/no-eol-whitespace/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (actual) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual });
    if (!validOptions) {
      return;
    }

    var rootString = root.toString();
    (0, _styleSearch2.default)({
      source: rootString,
      target: ["\n", "\r"],
      comments: "check"
    }, function (match) {
      if (whitespacesToReject.indexOf(rootString[match.startIndex - 1]) !== -1) {
        (0, _utils.report)({
          message: messages.rejected,
          node: root,
          index: match.startIndex - 1,
          result: result,
          ruleName: ruleName
        });
      }
    });
  };
};

var _utils = require("../../utils");

var _styleSearch = require("style-search");

var _styleSearch2 = _interopRequireDefault(_styleSearch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ruleName = exports.ruleName = "no-eol-whitespace";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: "Unexpected whitespace at end of line"
});

var whitespacesToReject = [" ", "\t"];
