// stylelint/dist/rules/function-url-quotes/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (expectation) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: expectation,
      possible: ["always", "never"]
    });
    if (!validOptions) {
      return;
    }

    root.walkAtRules(checkStatement);
    root.walkRules(checkStatement);

    function checkStatement(statement) {
      if (statement.type === "atrule") {
        checkAtRuleParams(statement);
      }

      statement.walkDecls(function (decl) {
        (0, _utils.functionArgumentsSearch)(decl.toString().toLowerCase(), "url", function (args, index) {
          checkArgs(args, decl, index, "url");
        });
      });
    }

    function checkAtRuleParams(atRule) {
      var atRuleParamsLowerCase = atRule.params.toLowerCase();
      (0, _utils.functionArgumentsSearch)(atRuleParamsLowerCase, "url", function (args, index) {
        checkArgs(args, atRule, index + (0, _utils.atRuleParamIndex)(atRule), "url");
      });
      (0, _utils.functionArgumentsSearch)(atRuleParamsLowerCase, "url-prefix", function (args, index) {
        checkArgs(args, atRule, index + (0, _utils.atRuleParamIndex)(atRule), "url-prefix");
      });
      (0, _utils.functionArgumentsSearch)(atRuleParamsLowerCase, "domain", function (args, index) {
        checkArgs(args, atRule, index + (0, _utils.atRuleParamIndex)(atRule), "domain");
      });
    }

    function checkArgs(args, node, index, functionName) {
      var leftTrimmedArgs = args.trimLeft();
      if (!(0, _utils.isStandardSyntaxUrl)(leftTrimmedArgs)) {
        return;
      }
      var complaintIndex = index + args.length - leftTrimmedArgs.length;
      var hasQuotes = leftTrimmedArgs[0] === "'" || leftTrimmedArgs[0] === "\"";
      switch (expectation) {
        case "always":
          if (hasQuotes) {
            return;
          }
          complain(messages.expected(functionName), node, complaintIndex);
          return;
        case "never":
          if (!hasQuotes) {
            return;
          }
          complain(messages.rejected(functionName), node, complaintIndex);
          return;
      }
    }

    function complain(message, node, index) {
      (0, _utils.report)({
        message: message,
        node: node,
        index: index,
        result: result,
        ruleName: ruleName
      });
    }
  };
};

var _utils = require("../../utils");

var ruleName = exports.ruleName = "function-url-quotes";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  expected: function expected() {
    var f = arguments.length <= 0 || arguments[0] === undefined ? "url" : arguments[0];
    return "Expected quotes around " + f + " argument";
  },
  rejected: function rejected() {
    var f = arguments.length <= 0 || arguments[0] === undefined ? "url" : arguments[0];
    return "Unexpected quotes around " + f + " argument";
  }
});
